compiler/luci/pass/src/CompressWeightsPass.cpp

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) 2024 Samsung Electronics Co., Ltd. All Rights Reserved
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "luci/Pass/CompressWeightsPass.h"
+#include "helpers/HuffmanEncoder.h"
+#include "helpers/NodeFiller.h"
+
+#include <luci/IR/CircleNodes.h>
+#include <luci/Service/Nodes/CircleConst.h>
+
+#include <cmath>
+#include <cassert>
+
+namespace
+{
+template <loco::DataType T> class TypeSelector;
+
+template <> class TypeSelector<loco::DataType::U8>
+{
+public:
+  using Type = uint8_t;
+};
+template <> class TypeSelector<loco::DataType::S8>
+{
+public:
+  using Type = int8_t;
+};
+
+template <loco::DataType DT> bool compress_weights_huffman_conv2d(luci::CircleConv2D *conv2d)
+{
+  using T = typename TypeSelector<DT>::Type;
+  assert(conv2d);
+  if (conv2d->weightCompression() == luci::WeightCompression::HUFFMAN)
+    return false;
+  luci::huffman::HuffmanEncoder<T> encoder;
+  auto weights = loco::must_cast<luci::CircleConst *>(conv2d->filter());
+  auto new_weights = luci::clone(weights);
+  std::vector<T> tmp_buf{};
+  for (size_t i = 0; i < weights->size<DT>(); ++i)
+  {
+    T data = weights->at<DT>(i);
+    tmp_buf.push_back(data);
+  }
+  std::vector<uint8_t> encoded = encoder.encode(tmp_buf);
+  new_weights->dtype(DT);
+  new_weights->size<DT>(encoded.size());
+  for (size_t i = 0; i < new_weights->size<DT>(); ++i)
+  {
+    new_weights->at<DT>(i) = encoded[i];
+  }
+  conv2d->filter(new_weights);
+  conv2d->weightCompression(luci::WeightCompression::HUFFMAN);
+
+  return true;
+}
+
+} // namespace
+
+namespace luci
+{
+
+bool CompressWeightsPass::run(loco::Graph *g)
+{
+  bool changed = false;
+  for (auto node : loco::active_nodes(loco::output_nodes(g)))
+  {
+    auto conv2d = dynamic_cast<luci::CircleConv2D *>(node);
+    if (not conv2d)
+      continue;
+    loco::DataType weights_dtype = loco::must_cast<luci::CircleConst *>(conv2d->filter())->dtype();
```

Comment:
please split lines. I'd like to have easy readable codes.

```suggestion
    auto filter = loco::must_cast<luci::CircleConst *>(conv2d->filter());
    auto weights_dtype = filter->dtype();
```